modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/sqlquery/DbExecutor.scala

```diff
@@ -83,24 +83,20 @@ object DbExecutor {
     new DbExecutor[F] {
       def getConnection(rdbms: Rdbms, connectionRef: ConnectionRef[F])(implicit M: Monad[F]): F[Either[Throwable, Connection]] =
         for {
-          cachedConnection <- connectionRef.get(()).map(flattenCached)
-          connection <- cachedConnection match {
-                          case Right(conn) =>
-                            for {
-                              closed <- Sync[F].delay(conn.isClosed)
-                              result <- if (!closed) conn.asRight[Throwable].pure[F]
-                                        else
-                                          for {
-                                            newConn <- Sync[F].delay(
-                                                         Either.catchNonFatal(DriverManager.getConnection(rdbms.connectionString))
-                                                       )
-                                            _ <- connectionRef.put((), newConn)
-                                          } yield newConn
-                            } yield result
-                          case Left(error) =>
-                            error.asLeft[Connection].pure[F]
-                        }
-        } yield connection
+          existingConn <- connectionRef.get(())
+          conn <- flattenCached(existingConn) match {
+                    case Right(conn) if !conn.isClosed =>
+                      Sync[F].pure(conn.asRight)
+                    // Connection closed or unitialized
+                    case Right(_) | Left(Unitialized) =>
+                      for {
+                        newConn <- Sync[F].delay(Either.catchNonFatal(DriverManager.getConnection(rdbms.connectionString)))
```

Comment:
https://github.com/snowplow/enrich/pull/481/commits/7293107af16d15bcd06d2f7a441d474a3cc04c70
I used [HikariCP](https://github.com/brettwooldridge/HikariCP)

To preserve the old behavior and not crash the app if for instance SQL credentials are wrong, I couldn't instantiate `HikariDataSource` already when creating `SqlQueryEnrichment` because it can throw a `SQLException`.

I considered going with a `Ref[F, Option[HikariDataSource]]` initialized to `None` but we can't use `Ref` with `Id` (it's missing `Sync`).

So I [used](https://github.com/snowplow/enrich/pull/481/commits/7293107af16d15bcd06d2f7a441d474a3cc04c70#diff-800f2303e20ced7e5977daa3b33212e2fb927cb94c7b1271fbf985ea315e10f6R121) a `lazy val` so that it's accessed for the first time only when enriching an event, and creating a bad row in case of problem.